### PR TITLE
Switch cloudant provider from cloudant library to ibmcloudant library

### LINKED ISSUE #21004: Switch from 'cloudant' to 'ibmcloudant'
### Apache Airflow Provider(s)

cloudant

### Apache Airflow version

main (development)

### What happened

The IBM Cloudant API client we’re currently using, [cloudant](https://pypi.org/project/cloudant/) (aka python-cloudant), reached EOL last December. The authors recommend switching to [ibmcloudant](https://pypi.org/project/ibmcloudant/). The switch itself doesn’t seem too complicated (disclaimer: I have zero knowledge on IBM Cloudant), but ibmcloudant is depending on `PyJWT<3.0.0,>=2.0.1`, which conflicts to Flask-Appbuilder’s `PyJWT<2.0.0,>=1.7.1`. So we need to unblock either pacakge somehow to make this work.

See also dpgaspar/Flask-AppBuilder#1691

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by topherinternational**
closes: #21004, cloudant is no longer developed and ibmcloudant is the active replacement

This is a breaking change to the provider in the following ways:

- `get_conn` now returns a `CloudantV1` object with different function names than the previous `Cloudant` object; the mappings from the old library functions to the new are outlined here: https://github.com/cloudant/python-cloudant/blob/master/MIGRATION.md#reference-table
- `get_conn` now directly returns a usable object instead of a context manager; `with`-block syntax is no longer needed
- The connection is now checked for a mandatory `host` value that functions as the Cloudant account name (previously the host field was optional, but could cause a runtime failure)

### CODE REVIEW

**Comment by topherinternational** [comment]
Rats, looks like the joke is on me and this PR is blocked on another third-party...`ibmcloudant` has a dependency conflict with the Snowflake Python library that nerfs the Airflow build for Python 3.8 and 3.9. 

There's an [issue](https://github.com/snowflakedb/snowflake-connector-python/issues/2016) for this in the Snowflake repo with an [open PR](https://github.com/snowflakedb/snowflake-connector-python/pull/1998), so hopefully it will be fixed soon and this Cloudant PR will unblock.

Ironic as this issue stalled for a while waiting for Flask-AppBuilder to fix a dependency pin.

Details:

`ibmcloudant` brings in `ibm-cloud-sdk-core` which [requires `urllib3` 2.x](https://github.com/IBM/python-sdk-core/blob/main/pyproject.toml#L30), which is fine for core Airflow, but the snowflake provider brings in `snowflake-connector-python`, which (for Python 3.8. and 3.9 only) requires urllib3 1.x.

**Comment by potiuk** [comment]
> Rats, looks like the joke is on me and this PR is blocked on another third-party...`ibmcloudant` has a dependency conflict with the Snowflake Python library that nerfs the Airflow build for Python 3.8 and 3.9.
> 
> There's an [issue](https://github.com/snowflakedb/snowflake-connector-python/issues/2016) for this in the Snowflake repo ([two issues actually](https://github.com/snowflakedb/snowflake-connector-python/issues/1997)) with an [open PR](https://github.com/snowflakedb/snowflake-connector-python/pull/1998), so hopefully it will be fixed soon and this Cloudant PR will unblock.
> 
> Ironic as this issue stalled for a while waiting for Flask-AppBuilder to fix a dependency pin.
> 
> Details:
> 
> `ibmcloudant` brings in `ibm-cloud-sdk-core` which [requires `urllib3` 2.x](https://github.com/IBM/python-sdk-core/blob/main/pyproject.toml#L30), which is fine for core Airflow, but the snowflake provider brings in `snowflake-connector-python`, which (for Python 3.8. and 3.9 only) requires urllib3 1.x.

You could potentially also add "3.8" and "3.9" to "excluded-python-versions" in the new version of cloudant provider (in provider.yaml) with the link to the snowflake issue - and remove the exclusions once this problem is fixed.

**Comment by potiuk** [comment]
See https://github.com/apache/airflow/pull/41548 - where we just removed such exclusion for Papermill for Python 3.12

**Comment by potiuk** [comment]
I've added "all-versions" tag to the PR of yours - to make sure it's tested on all versions - so once you exclude python versions you should be able to see both CI images and tests to run for cloudant only on 3.10 -> 3.12 images and be skipped for 3.8 and 3.9

**Comment by topherinternational** [comment]
Great explanation, thank you @potiuk

**Comment by topherinternational** [comment]
Ok @potiuk I've added the exclusion and an explanatory comment, so I think this is complete. Please review at your convenience, thanks.

**Comment by potiuk** [comment]
You should also add a note at the top of changelog (see comment at the top of changelog) and bump the version to the new major release in the provider.yaml.

File: airflow/providers/cloudant/provider.yaml
Comment (eladkal) [inline airflow/providers/cloudant/provider.yaml:1]: Please `4.0.0` to versions. Since we know for sure this PR is breaking

```
versions:
  - 4.0.0
```

File: airflow/providers/cloudant/provider.yaml
Comment (eladkal) [inline airflow/providers/cloudant/provider.yaml]: Can you clarify about this?
I see it does: https://github.com/IBM/python-sdk-core/blob/d603ed0476fb3ceca0b1e82ae1259d3725e2948f/pyproject.toml#L17

Comment (topherinternational) [reply airflow/providers/cloudant/provider.yaml]: Yes I can clarify - 3.12 support was added to ibm-cloud-sdk-core in 3.20.5, the latest ibmcloundant pins ibm-cloud-sdk-core at 3.20.3. (I don't think anything actually changed in the code or deps, they just added 3.12 to the classifiers.)

Comment (eladkal) [reply airflow/providers/cloudant/provider.yaml]: Is there an issue or pr we can track progress on this?

Comment (topherinternational) [reply airflow/providers/cloudant/provider.yaml]: ibmcloudant has already [bumped ibm-cloud-sdk-core to 3.20.6](https://github.com/IBM/cloudant-python-sdk/commit/ff052f7ce1280d53954dbd07a62f8f580fc7f65a), however they haven't released yet. I can't find an official release schedule in the repo, but their releases look to be timed 4-8 weeks apart so a new one should be coming soon.

Would you like me to file an issue in the Airflow project or is a note/TODO in the provider.yaml sufficient?

File: docs/spelling_wordlist.txt
Comment (topherinternational) [inline docs/spelling_wordlist.txt:254]: When I tried to add `CloudantV1` to the wordlist the check still failed claiming `CloundantV` wasn't in the list. The actual class name is `CloudantV1`, but the spell checker seems to interpret the string as `CloudantV` with the number 1 appended.

Comment (potiuk) [reply docs/spelling_wordlist.txt:254]: Add `` around. (`CloundantV1`)

File: airflow/providers/cloudant/provider.yaml
Comment (topherinternational) [inline airflow/providers/cloudant/provider.yaml:28]: Some provider tests were failing with error messages about 3.6.0 not being a valid revision, trace below. We have a tag for providers-cloudant/3.6.0rc1 but not for 3.6.0 GA; is it legit to change the newest version to 4.0.0, will that make the release process work?

```
New version of the 'cloudant' package is ready to be released!
Running command: <git log apache-https-for-providers/main...providers-cloudant/3.6.0 .>
Working directory <project-dir>/airflow/airflow/providers/cloudant

git log '--pretty=format:%H %h %cd %s' --date=short apache-https-for-providers/main...providers-cloudant/3.6.0 -- .

========================= STDERR start ============================
fatal: bad revision 'apache-https-for-providers/main...providers-cloudant/3.6.0'

```

Comment (potiuk) [reply airflow/providers/cloudant/provider.yaml:28]: It will have to wait for release of 3.6.0 to happen

Comment (potiuk) [reply airflow/providers/cloudant/provider.yaml:28]: (2 days or so)

File: airflow/providers/cloudant/cloudant_fake.py
Comment (topherinternational) [inline airflow/providers/cloudant/cloudant_fake.py:1]: @eladkal @potiuk so this is a big hack...static tests were failing bc we run static tests on Pyhton 3.8, but we exclude the provider on 3.8, so the container doesn't have the ibmcloudant package installed and the module fails to evaluate. So I added a version detection block that imports a fake version of the couldant classes on 3.8 and 3.9. 

I leave it to you to decide if you want to have this kind of mess in your codebase 😅 . I'm not sure if exclusions of the lowest-supported Python version is feasible without something like this to keep the code from being eval'd at all in the static tests and doc tasks.

Comment (potiuk) [reply airflow/providers/cloudant/cloudant_fake.py:1]: We should just leave a note  / comment to drop it after the issue is solved ( PR merged)  explaining why it is added and link to the PR/issue.

File: airflow/providers/cloudant/provider.yaml
Comment (eladkal) [reply airflow/providers/cloudant/provider.yaml]: Todo is needed any way. The issue is that it should have something simple to track (issue/pr) we should avoid leaveing exploration tasks. I think filing an issue is simple enough and will provide the clarity we need

**Comment by topherinternational** [comment]
@potiuk @eladkal we have 3 persistent failures in the provider checks; tldr, I think there's a bug-ish in the provider checks commands in breeze where all providers are assumed to be supported on Python 3.8. 

The command `breeze release-management verify-provider-packages --use-packages-from-dist --package-format wheel --use-airflow-version wheel --airflow-constraints-reference default --providers-constraints-location /files/constraints-3.8/constraints-source-providers-3.8.txt` installs all of the provider wheels built in the earlier `breeze release-management prepare-provider-packages ...` step. 

This process all runs on 3.8 and so implicitly assumes that all of the providers are legit for 3.8; neither step looks like it checks `excluded-python-versions`. So the install fails due to the ibmcloudant and snowflake dependency conflict (the very reason we excluded 3.8 in the first place). 

These commands happen in `Provider packages wheel build and verify`, `Compat 2.8.4:P3.8 provider check` and `Compat 2.9.1:P3.8 provider check`.

Addressing this is obviously your call, but maybe some options are?:
1. Make `prepare-provider-packages` check excluded versions before building a particular provider wheel
2. Add "Remove incompatible <version> provider packages" task to `Provider packages wheel build and verify`, and have the task reference the provider dependencies json and the excluded python versions.
3. Run the provider checks on every supported python version (after doing 1 or 2)

I checked the repo and none of the released providers exclude Python 3.8, I figure maybe this problem hasn't come up yet.

**Comment by topherinternational** [comment]
We also have spurious failures where one of the tests in "Lowest direct dependency resolution tests" and one in "Kubernetes tests" will fail, but it looks like a different one each time. Suspiciously the failures take about twice as long as any of the successes (2hrs vs ~55mins) so maybe a resource starvation or a hang in a testing resource.

**Comment by potiuk** [comment]
> We also have spurious failures where one of the tests in "Lowest direct dependency resolution tests" and one in "Kubernetes tests" will fail, but it looks like a different one each time. Suspiciously the failures take about twice as long as any of the successes (2hrs vs ~55mins) so maybe a resource starvation or a hang in a testing resource.

Yes. Some flakes.

**Comment by potiuk** [comment]
> I checked the repo and none of the released providers exclude Python 3.8, I figure maybe this problem hasn't come up yet.

Yes. As I wrote in the #41718 - likely matrix for all providers and python versions is far too big an overhead for this one - this is the first time we have a provider that does not work for 3.8. So likely removing the providers that are not 3.8 compatible is the way to go - feel free to do so. We already have a code that removes non-compatible providers in compatibility checks - you can add ibmcloudant there (with comment referring to the conflict and issue with snowflake). Similarly you can even hard-code a step after builiding the providers in verify providers (again - with comment and link to the snowflake issue and note that it should be removed after the issue is solved.

BASE_PROVIDERS_COMPATIBILITY_CHECKS: list[dict[str, str | list[str]]] = [
    {
        "python-version": "3.8",
        "airflow-version": "2.8.4",
        "remove-providers": "fab",
        "run-tests": "true",
    },
    {
        "python-version": "3.8",
        "airflow-version": "2.9.3",
        "remove-providers": "",
        "run-tests": "true",
    },
    {
        "python-version": "3.8",
        "airflow-version": "2.10.0",
        "remove-providers": "",
        "run-tests": "true",
    },
]

File: airflow/providers/cloudant/provider.yaml
Comment (eladkal) [inline airflow/providers/cloudant/provider.yaml]:
```suggestion
  - 4.0.0
  - 3.6.0
  - 3.5.2
```

@topherinternational you have wrong edit here. you removed 3.6.0, please restore it.

**Comment by potiuk** [comment]
There is one more issue - when package is excluded fromcertai version , the 'lowest deendencies' test goes   wild - it tests all other packages then. This should be fixed as part of CI infra fix

**Comment by potiuk** [comment]
I added a fixup that is supposed to skip cloundant tests for 3.8, 3.9, 3.12 @topherinternational

**Comment by potiuk** [comment]
Almost there. For some strange reason apache.beam "downgrade" does not work quite consistently and hangs for Python 3.12.

 Need to take a look

**Comment by potiuk** [comment]
I guess we will also have to exclude Python 3.12 version - until apache.beam fixes the conflicts.

**Comment by topherinternational** [comment]
> There is one more issue - when package is excluded fromcertai version , the 'lowest deendencies' test goes   wild - it tests all other packages then. This should be fixed as part of CI infra fix

I found the bug in breeze that causes this, I am coding it up into a separate PR.

**Comment by potiuk** [comment]
Hmm. The problem is this for apache.beam

```
--verbosity=0 --strict-markers --durations=100 --maxfail=50 --color=yes --junitxml=/files/test_result-providers_apache_beam-postgres.xml --timeouts-order moi --setup-timeout=60 --execution-timeout=60 --teardown-timeout=60 --disable-warnings -rfEX --ignore=tests/system --ignore=tests/integration --warning-output-path=/files/warnings-providers_apache_beam-postgres.txt --ignore=helm_tests --with-db-init --ignore=tests/providers/apache/beam --ignore=tests/system/providers/apache/beam --ignore=tests/integration/providers/apache/beam --no-cov

```

It runs `all other` tests as beam is also excluded for Python 3.12

**Comment by potiuk** [comment]
This one should fix it https://github.com/apache/airflow/pull/41991

**Comment by topherinternational** [comment]
> Hmm. The problem is this for apache.beam
> 
> 
> 
> ```
> 
> --verbosity=0 --strict-markers --durations=100 --maxfail=50 --color=yes
> 
> --junitxml=/files/test_result-providers_apache_beam-postgres.xml
> 
> --timeouts-order moi --setup-timeout=60 --execution-timeout=60 --teardown-timeout=60 
> 
> --disable-warnings -rfEX --ignore=tests/system --ignore=tests/integration
> 
> --warning-output-path=/files/warnings-providers_apache_beam-postgres.txt
> 
> --ignore=helm_tests --with-db-init --ignore=tests/providers/apache/beam
> 
> --ignore=tests/system/providers/apache/beam
> 
> --ignore=tests/integration/providers/apache/beam --no-cov
> 
> ```
> 
> 
> 
> It runs `all other` tests as beam is also excluded for Python 3.12

Hmm, that looks like the same problem I brought up in #41967, Breeze removes the test dir argument for the excluded provider, and so the command winds up running all of the tests.

**Comment by potiuk** [comment]
I rebased to see if the fix #41991 is merged

**Comment by topherinternational** [comment]
Opened #42027 which I think fixes the current test failures.

**Comment by potiuk** [comment]
🤞 🤞 🤞 🤞 🤞 🤞 🤞 🤞

**Comment by topherinternational** [comment]
Another cross-fix: #42030

**Comment by potiuk** [comment]
🤞 🤞 🤞 🤞 🤞 🤞 🤞 🤞🤞 🤞 🤞 🤞 🤞 🤞 🤞 🤞

File: airflow/providers/cloudant/provider.yaml
Comment (topherinternational) [reply airflow/providers/cloudant/provider.yaml]: I removed the 3.12 exclusion as I think the test failures were from apache beam and other issues, not the ibmcloudant dependencies. If tests pass we know it's gtg on 3.12.

File: dev/breeze/src/airflow_breeze/utils/selective_checks.py
Comment (topherinternational) [inline dev/breeze/src/airflow_breeze/utils/selective_checks.py:1281]: Python makes this more complicated than it could be but I figured this dictionary should be sorted by Python version, esp since our test for this is a string equivalency and not just `dict == dict`.

**Comment by potiuk** [comment]
🤞 🤞 🤞 🤞 🤞 🤞 🤞 🤞🤞 🤞 🤞 🤞 🤞 🤞 🤞 🤞🤞 🤞 🤞 🤞 🤞 🤞 🤞 🤞🤞 🤞 🤞 🤞 🤞 🤞 🤞 🤞

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
Woooooohoooo!

**Comment by topherinternational** [comment]
> Woooooohoooo!

This has been...educational 😉

**Comment by potiuk** [comment]
> This has been...educational 

Oh, I am sure it was :)

**Pull request merged**
